include/zephyr/spinlock.h

```diff
@@ -29,6 +29,35 @@ extern "C" {
  * @{
  */
 
+/**
+ * @brief Execute a code block with the given spinlock held, automatically
+ * releasing it in the end.
+ *
+ * @details Example usage:
+ *
+ * @code{.c}
+ * WITH_SPINLOCK(&mylock) {
+ *   ...execute protected statements with the lock held...
+ * }
+ * @endcode
+ *
+ * This basically expands to:
+ *
+ * @code{.c}
+ * for (k_spinlock_key_t key = k_spin_lock(&mylock); k_spin_unlock(&mylock, key); ...) {
+ *     ...execute protected statements with the lock held...
+ * }
+ * @endcode
+ *
+ * @note This pattern may only be used, if you need to protect a critical
+ * section with a single point of exit in the end. As it uses a for-loop
+ * internally, you MUST NOT EXIT ANYWHERE FROM THE BLOCK (e.g. using a break or
+ * return statement) otherwise the spinlock will not be released.
+ */
+#define WITH_SPINLOCK(lck)                                                                         \
```

Comment:
Possible name change? K_HOLD_SPINLOCK ?

Reply:
`SPINLOCK_GUARDED`? Since you are creating a new macro, it is a good idea to make it more descriptive.

Reply:
@peter-mitsis @dcpleung K_SPINLOCK_GUARDED maybe? I think the K_ prefix makes sense here.